lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,348 @@
+      newOperands.push_back(value);
+      // Prepare the additional input ports that need to be added to the
+      // referenced module.
+      newInputs.emplace_back(name, value.getType(), value.getLoc());
+    }
+    // Add the new operands to the instance.
+    instance->insertOperands(instance->getNumOperands(), newOperands);
+    if (!argNames.empty())
+      instance->setAttr("argNames",
+                        ArrayAttr::get(instance->getContext(), argNames));
+
+    // Now update other instances of the module.
+    for (auto otherInstance : moduleInstances) {
+      if (otherInstance == instance)
+        continue;
+      SmallVector<Value, 2> otherInstanceNewOperands;
+      otherInstanceNewOperands.reserve(inputs.size());
+      mlir::OpBuilder builder(otherInstance);
+      for (unsigned i = 0; i < inputs.size(); ++i)
+        otherInstanceNewOperands.emplace_back(builder.create<ConstantOp>(
+            otherInstance->getLoc(), builder.getI1Type(), 1));
+
+      otherInstance->insertOperands(otherInstance->getNumOperands(),
+                                    otherInstanceNewOperands);
+      if (otherInstance->hasAttr("argNames"))
+        otherInstance->setAttr(
+            "argNames", ArrayAttr::get(instance->getContext(), argNames));
+    }
+    // Update all the modules with the new input ports.
+    refMod.appendPorts(newInputs, {});
+  }
+
+  /// Get the referenced module for an instance, and add all the instances of
+  /// the module to `instances`.
+  HWModuleOp getReferencedModules(HWInstanceLike inst,
+                                  SmallVector<HWInstanceLike, 2> &instances) {
+    HWModuleOp instModule = {};
+    auto refModNames = inst.getReferencedModuleNamesAttr();
+    if (refModNames.size() != 1) {
+      inst.emitWarning("expected an instance with a single reference");
+      return {};
+    }
+    auto *node = graph->lookup(cast<StringAttr>(*refModNames.begin()));
+    if (node)
+      instModule = dyn_cast_or_null<HWModuleOp>(*node->getModule());
+    if (instModule.isPublic())
+      return {};
+
+    for (auto *use : node->uses()) {
+      auto otherInstance = use->getInstance<HWInstanceLike>();
+      if (!otherInstance) {
+        instModule.emitWarning("Cannot find all instances");
+        return {};
+      }
+      if (otherInstance.getReferencedModuleNamesAttr().size() != 1) {
+
+        otherInstance.emitWarning(
+            "expected an instance with a single reference");
+        return {};
+      }
+      instances.push_back(otherInstance);
+    }
+
+    return instModule;
+  }
+
+  /// If the `argNames` attribute exists, return it.
+  SmallVector<StringAttr> getInstanceArgNames(HWInstanceLike inst) {
+    SmallVector<StringAttr> argNames;
+    if (auto args = inst->getAttrOfType<ArrayAttr>("argNames")) {
+      auto attr = args.getAsRange<StringAttr>();
+      argNames.insert(argNames.begin(), attr.begin(), attr.end());
+    }
+    return argNames;
+  }
```

Comment:
Could you add `argNames` getter to HWInstanceLike. I guess all operations that implement HWInstanceLike have non-optional `argNames`